### PR TITLE
[image_picker] Image picker fix camera device

### PR DESCRIPTION
**Pull request by ydag**
This PR fixes an issue where `preferredCameraDevice` is not working as expected for `getVideo` method. I also refactor device-only unit tests to make sure we can test camera devices on the CI as well.

Related issues :
- Fixes [flutter/flutter#74674](https://github.com/flutter/flutter/issues/74674)

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`. See [plugin_tool format](../script/tool/README.md#format-code))
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

File: packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m
Comment (stuartmorgan) [inline packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:464]: What is this change for? It doesn't seem obviously related.

File: packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m
Comment (stuartmorgan) [inline packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:164]: This mapping from a magic number, as well as the hard-coded key, is now duplicated in two places. Please extract a small helper method that takes an arguments dictionary and returns the device, so that both locations can share the same code.

**Review by stuartmorgan — COMMENTED** [review]
We should have tests covering this. Long-term that will probably require some refactoring for testability, but short term: was [this test](https://github.com/flutter/plugins/blob/master/packages/image_picker/image_picker/ios/Tests/ImagePickerPluginTests.m#L98) failing and we just didn't know because it's device-only?

File: packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m
Comment (ydag) [reply packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:464]: Yes, this is not related to this PR but since it is just a small fix I wanted to add it here. This change is related to [#3873](https://github.com/flutter/plugins/pull/3873). `isMetadataAvailable` parameter is used to handle rotation problem for limited access to the photo library and it should be `YES` for all other occasions.

File: packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m
Comment (ydag) [reply packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:164]: done.

**Comment by ydag** [comment]
> We should have tests covering this. Long-term that will probably require some refactoring for testability, but short term: was [this test](https://github.com/flutter/plugins/blob/master/packages/image_picker/image_picker/ios/Tests/ImagePickerPluginTests.m#L98) failing and we just didn't know because it's device-only?

I think the problem is not the test. Since `UIImagePickerControllerCameraDevice _device` is a global variable we need to re-assign before we call the `showCamera` method from both `pickImage` and `pickVideo` in `handleMethodCall`.

But I double-check and I think we even don't need a `UIImagePickerControllerCameraDevice _device` variable at all if we call the `getCameraDevice` helper method inside the `showCamera` method.

File: packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m
Comment (stuartmorgan) [inline packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:191]: Please use a local variable instead of calling this twice.

File: packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m
Comment (stuartmorgan) [inline packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:166]: Why this change?

File: packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m
Comment (stuartmorgan) [reply packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:466]: If it's not related, then it would be much better to create a new PR for it. Having behavior-changing fixes hidden in unrelated PRs makes things much more difficult when doing things like trying to find the reason for certain changes with blame, or looking for causes of regressions.

It will also need tests, which would require adding even more (unrelated) code here.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
> > We should have tests covering this. Long-term that will probably require some refactoring for testability, but short term: was [this test](https://github.com/flutter/plugins/blob/master/packages/image_picker/image_picker/ios/Tests/ImagePickerPluginTests.m#L98) failing and we just didn't know because it's device-only?
> 
> I think the problem is not [this test](https://github.com/flutter/plugins/blob/master/packages/image_picker/image_picker/ios/Tests/ImagePickerPluginTests.m#L98). Since `UIImagePickerControllerCameraDevice _device` is a global variable we need to re-assign before we call the `showCamera` method from both `pickImage` and `pickVideo` in `handleMethodCall`.

Sorry, I'm not seeing how this answers my question. To clarify, there are two parts:
- Where are the test changes that this PR needs to verify the fix, per Flutter policy?
- What is the status of the test referenced above, which sounds like it should already have been testing this scenario, but had not alerted us to the failure previously given that the tree is green?

File: packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m
Comment (ydag) [reply packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:166]: I am not sure why I changed this. I reverted it.

Comment (ydag) [reply packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:166]: I am not sure why I changed this. I reverted it.

File: packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m
Comment (ydag) [reply packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:191]: done.

File: packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m
Comment (ydag) [reply packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:69]: Done.

**Comment by stuartmorgan** [comment]
That means we don't have any meaningful test coverage here, since a regression would (as evidenced by this PR) never show up in our CI.

It looks like this should be testable via native unit tests. It would require class mocking of `AVCaptureDevice` and `UIImagePickerController` unfortunately, since there are class methods called on them, but class mocks are still better than no test coverage at all. Alternately, you could wrap those calls in a helper that can be instantiated and mocked less invasively via DI.

File: packages/image_picker/image_picker/CHANGELOG.md
Comment (cyanglaz) [inline packages/image_picker/image_picker/CHANGELOG.md:2]:
```suggestion
## 0.8.0+2

* Fixes an issue where `preferredCameraDevice` option is not working for `getVideo` method.
```

File: packages/image_picker/image_picker/example/ios/Podfile
Comment (cyanglaz) [inline packages/image_picker/image_picker/example/ios/Podfile:37]: The target of the unit tests have been changed to `RunnerTests` recently, so this needs to be updated.

File: packages/image_picker/image_picker/CHANGELOG.md
Comment (ydag) [reply packages/image_picker/image_picker/CHANGELOG.md]: Done.

File: packages/image_picker/image_picker/example/ios/Podfile
Comment (ydag) [reply packages/image_picker/image_picker/example/ios/Podfile:37]: Done.

File: packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m
Comment (cyanglaz) [inline packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m:70]: This comment seems not necessary, the code below is self-explanatory.

File: packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m
Comment (cyanglaz) [inline packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m:98]: This comment seems not necessary, the code below is self-explanatory.

File: packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m
Comment (cyanglaz) [inline packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m:126]: This comment seems not necessary, the code below is self-explanatory.

File: packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m
Comment (cyanglaz) [inline packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m:154]: This comment seems not necessary, the code below is self-explanatory.

**Review by cyanglaz — APPROVED** [review]
LGTM % nits

**Comment by mvanbeusekom** [comment]
> That means we don't have any meaningful test coverage here, since a regression would (as evidenced by this PR) never show up in our CI.
> 
> It looks like this should be testable via native unit tests. It would require class mocking of `AVCaptureDevice` and `UIImagePickerController` unfortunately, since there are class methods called on them, but class mocks are still better than no test coverage at all. Alternately, you could wrap those calls in a helper that can be instantiated and mocked less invasively via DI.

@stuartmorgan could you have another look at this PR? @ydag updated the tests so the device specific features are mocked using OCMock and unit-tests will run correctly on the simulator. Chris already approved but the PR is blocked on your requested changes.

File: packages/image_picker/image_picker/example/ios/Podfile
Comment (stuartmorgan) [inline packages/image_picker/image_picker/example/ios/Podfile:34]: Why are you setting the tests to iOS 9? We still support iOS 8 for now. (The messages about deployment target when building with recent versions of Xcode should be harmless.)

File: packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m
Comment (stuartmorgan) [inline packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m:44]: If you're not using tearDown you can remove it.

File: packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m
Comment (stuartmorgan) [inline packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m:51]: The true value for a BOOL is `YES`, not `true`.

Applies throughout the PR.

File: packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m
Comment (stuartmorgan) [inline packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m:59]: Why isn't this returning an `AVAuthorizationStatus` constant?

File: packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m
Comment (stuartmorgan) [inline packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m:155]: Blank line after `pragma mark` please.

File: packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m
Comment (stuartmorgan) [inline packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m:157]: Why are you aborting the test if the camera *is* available?

File: packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m
Comment (stuartmorgan) [inline packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m:171]: Why is this called twice?

File: packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m
Comment (stuartmorgan) [inline packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m:172]: Why isn't this test asserting anything?

File: packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m
Comment (stuartmorgan) [inline packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:79]: `arguments` is the name of the parameter, not a Doxygen command, so should not have a `@`.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/image_picker/image_picker/example/ios/Podfile
Comment (ydag) [reply packages/image_picker/image_picker/example/ios/Podfile:34]: I set it because `OCMock` requires iOS 9.

File: packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m
Comment (ydag) [reply packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m:157]: Hi Stuart,
this change made by [this commit](https://github.com/flutter/plugins/commit/25a585dfeef6d2dc224fd0dddee327372de63782#) and it already exists in master as well and got merged in my PR. Do you think we should remove it? Or ask what was this test about and try to fix it?

Comment (stuartmorgan) [reply packages/image_picker/image_picker/example/ios/RunnerTests/ImagePickerPluginTests.m:157]: @cyanglaz Was this a bug, or is this intentionally only supposed to run on simulator? That seems bizarre.

File: packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m
Comment (stuartmorgan) [reply packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:79]: This hasn't been changed.

Comment (ydag) [reply packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:79]: Oops, sorry. I also changed other comments with the same mistake. :)

File: packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m
Comment (stuartmorgan) [inline packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:73]: It doesn't though; it returns the device that should be used. How about "Returns the UIImagePickerControllerCameraDevice to use given [arguments]."

**Review by stuartmorgan — APPROVED** [review]
LGTM with one comment nit (no need to get another review once fixed)

**Pull request merged**
